File: src/1102-Path-With-Maximum-Minimum-Value/1102.cpp
class Solution 
{
public:
    int maximumMinimumPath(vector<vector<int>>& A) 
    {
        int r = A.size(), c = A[0].size();
        vector<vector<int>> vi(r, vector<int>(c, 0));
        priority_queue<vector<int>> q;
        q.push({A[0][0], 0, 0});
        while (!q.empty())
        {
            auto it = q.top(); q.pop();
            if (it[1] == r-1 and it[2] == c-1) return it[0];
            for (auto& d : dire)
            {
                int nx = d[0] + it[1], ny = d[1] + it[2];
                if (nx >= 0 and nx < r and ny >= 0 and ny < c
                    and vi[nx][ny] == 0)
                {
                    vi[nx][ny] = 1;
                    q.push({min(it[0], A[nx][ny]), nx, ny});
                }
            }
        }
        return 0;
    }
private:
    int dire[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
};

File: src/1102-Path-With-Maximum-Minimum-Value/1102.go
import "container/heap"
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

type Item struct {
    pre, x, y int
}

type Heap []*Item

func (h Heap) Len() int { 
    return len(h)
}

func (h Heap) Less(i, j int) bool {
    return h[i].pre > h[j].pre
}

func (h Heap) Swap(i, j int) {
    h[i], h[j] = h[j], h[i]
}

func (h *Heap) Pop() interface{} {
    old := *h
    n := len(old)
    item := old[n - 1]
    *h = old[0:n - 1]
    return item
}

func (h *Heap) Push(x interface{}) {
    *h = append(*h, x.(*Item))
}

func maximumMinimumPath(A [][]int) int {
    r, c := len(A), len(A[0])
    dire := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
    vi := make([][]int, r)
    for i := 0; i < r; i++ {
        vi[i] = make([]int, c)
    }
    vi[0][0] = 1
    
    q := &Heap{&Item{A[0][0], 0, 0}}
    heap.Init(q)
    for q.Len() > 0 {
        it := heap.Pop(q).(*Item)
        if it.x == r-1 && it.y == c-1 {
            return it.pre
        }
        for _, v := range dire {
            nx, ny := it.x + v[0], it.y + v[1]
            if nx >= 0 && nx < r && ny >= 0 && ny < c && vi[nx][ny] == 0 {
                vi[nx][ny] = 1
                heap.Push(q, &Item{min(it.pre, A[nx][ny]), nx, ny})
            } 
        }
    }
    return 0
}
